include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+    }
+
+    [[nodiscard]] inline bool equals(const ::in6_addr& lhs, const ::in6_addr& rhs) WI_NOEXCEPT
+    {
+        return 0 == ::memcmp(&lhs, &rhs, sizeof(::in6_addr));
+    }
+
+    [[nodiscard]] inline bool not_equals(const ::in6_addr& lhs, const ::in6_addr& rhs) WI_NOEXCEPT
+    {
+        return !::wil::network::equals(lhs, rhs);
+    }
+
+    //
+    // encapsulates working with the sockaddr datatype
+    //
+    // sockaddr is a generic type - similar to a base class, but designed for C with BSD sockets (1983-ish)
+    // 'derived' structures are cast back to sockaddr* (so the initial struct members must be aligned)
+    //
+    // this data type was built to be 'extensible' so new network types could create their own address structures
+    // - appending fields to the initial fields of the sockaddr
+    //
+    // note that the address and port fields of TCPIP sockaddr* types were designed to be encoded in network-byte order
+    // - hence the common use of "host-to-network" and "network-to-host" APIs, e.g. htons(), htonl(), ntohs(), ntohl()
+    //
+    // Socket APIs that accept a socket address will accept 2 fields:
+    // - the sockaddr* (the address of the derived sockaddr type, cast down to a sockaddr*)
+    // - the length of the 'derived' socket address structure referenced by the sockaddr*
+    //
+    // Commonly used sockaddr* types that are using with TCPIP networking:
+    //
+    // sockaddr_storage / SOCKADDR_STORAGE
+    //   - a sockaddr* derived type that is guaranteed to be large enough to hold any possible socket address (not just TCPIP related)
+    // sockaddr_in / SOCKADDR_IN
+    //   - a sockaddr* derived type designed to contain an IPv4 address and port number
+    // sockaddr_in6 / SOCKADDR_IN6
+    //   - a sockaddr* derived type designed to contain an IPv6 address, port, scope id, and flow info
+    // SOCKADDR_INET
+    //   - a union of sockaddr_in and sockaddr_in6 -- i.e., large enough to contain any TCPIP IPv4 or IPV6 address
+    // in_addr / IN_ADDR
+    //   - the raw address portion of a sockaddr_in
+    // in6_addr / IN6_ADDR
+    //   - the raw address portion of a sockaddr_in6
+    //
+    // SOCKET_ADDRESS
+    //   - not a derived sockaddr* type
+    //   - a structure containing both a sockaddr* and its length fields, returned from some networking functions
+    //
+
+    // declaring char-arrays large enough for any IPv4 or IPv6 address + optional fields
+    static_assert(INET6_ADDRSTRLEN > INET_ADDRSTRLEN);
+    typedef WCHAR socket_address_wstring[INET6_ADDRSTRLEN];
+    typedef CHAR socket_address_string[INET6_ADDRSTRLEN];
+
+    class socket_address final
+    {
+    public:
+        constexpr socket_address() WI_NOEXCEPT = default;
+        explicit socket_address(ADDRESS_FAMILY family) WI_NOEXCEPT;
+        template <typename T>
+        explicit socket_address(_In_reads_bytes_(addr_size) const SOCKADDR* addr, T addr_size) WI_NOEXCEPT;

```

Comment:
Does this need to be a template? Just pass as `size_t`

Reply:
I made it a template since I wanted to avoid unnecessary casting that was likely going to be happening all over.

Reply:
The downside to this approach is that this code is now responsible for proper casting. What if someone calls with a negative signed integer? Or a value larger than a `size_t` can hold? Also, I'm hoping callers are using `sizeof` when calling, which already has type `size_t`

Reply:
it's handled in the eventual function + they'll get a WI_ASSERT.

    template <typename T>
    void socket_address::reset(_In_reads_bytes_(addr_size) const SOCKADDR* addr, T addr_size) WI_NOEXCEPT
    {
        WI_ASSERT(static_cast<size_t>(addr_size) <= static_cast<size_t>(size()));

        ::memset(&m_sockaddr, 0, size());
        if (addr)
        {
            ::memcpy_s(&m_sockaddr, sizeof(m_sockaddr), addr, addr_size);
        }
    }

Reply:
Do you have any example code that would use this constructor/`reset` function, but not use something like `sizeof`? In the tests, I only see two calls to the constructor, both of which use `sizeof` (though both also use either `SOCKADDR_IN` or `SOCKADDR_IN6`, each with their own constructor).

There's also a call to `reset` in the iterator type, however `ai_addrlen` is defined as a `size_t` in all of `ADDRINFOA`, `ADDRINFOW`, and `ADDRINFOEXW`.

Reply:
Yep. You're right. This can and should be size_t.
Fixing in the next update.